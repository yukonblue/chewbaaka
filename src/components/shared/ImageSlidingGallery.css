/**
 * ImageSlidingGallery.css
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 06, 2020
 * Updated  : Jun 09, 2022
 */

div.ImageSlidingGalleryCoreContainer {
  position: relative;
  padding-left: 55px;
}

div.ImageSlidingGalleryButtonContainer {
  position: absolute;
  justify-content: center;
  margin: 0;
  left: 0px;
  top: 50%;
  transform: translateY(-50%);
}
