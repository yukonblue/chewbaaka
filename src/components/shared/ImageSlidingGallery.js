/**
 * ImageSlidingGallery.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 06, 2020
 * Updated  : Jun 15, 2022
 */

/**
 * ImageSlidingGallery
 *
 * `ImageSlidingGallery` is a component that illustrates a set of
 * user specified images one at a time, along with a piece of
 * caption string annotating the image.
 *
 * It is advisable to have all the images in the same dimension,
 * or at least same aspect ratios. Otherwise the rendering output
 * may cause "page jump" effect.
 *
 * https://github.com/tetrachrome/chewbaaka/issues/95
 *
 * Props:
 *
 *  - `slides`: An array of objects containing the image and caption
 *    tuples.
 */

import React from 'react'

import { Button, Transition } from "semantic-ui-react"

import uniqid from 'uniqid'

import 'semantic-ui-css/semantic.min.css'

import { getElementStyleClassName } from '../../styling/styling'

import Caption from './Caption'

import { getFormattedImageCaptionStringWithCredit } from './ImageCaptionUtils'

import {
  Design2ImageViewFamilyImgStyles,
  Design2ImageViewFamilyContainerStyles,
  Design2ImageViewFamilyCaptionContainerStyles
} from './Design2_ImageViewFamilyUtils'

import './ImageSlidingGallery.css'

import './Design2.css'

if ( process.env.NODE_ENV === 'development' )
  require('./ImageSlidingGallery-debug.css')

export default class ImageSlidingGallery extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      activeIndex : 0
    }
    this.handleOnPrevClick = this.handleOnPrevClick.bind(this);
    this.handleOnNextClick = this.handleOnNextClick.bind(this);
  }

  handleOnPrevClick() {
    this.setState((prevState) => ({
      activeIndex : Math.max(prevState.activeIndex-1, 0)
    }));
  }

  handleOnNextClick() {
    this.setState((prevState) => ({
      activeIndex : Math.min(prevState.activeIndex+1, this.props.slides.length - 1)
    }));
  }

  render() {
    const componentOuterContainerDimension = {
      maxWidth: this.props.width
    };

    const activeSlide = this.props.slides[this.state.activeIndex];

    const caption = getFormattedImageCaptionStringWithCredit(activeSlide.caption, activeSlide.credit);

    const aspectRatio = activeSlide.aspectRatio;

    return (
      <div
        className={getElementStyleClassName("ImageSlidingGalleryOuterContainer")}
        style={componentOuterContainerDimension}
      >
        <div className="ImageSlidingGalleryCoreContainer">
          <div className={Design2ImageViewFamilyContainerStyles}>
            <Transition key={uniqid()} transitionOnMount={true} visible={true} duration={500} animation="fade right">
              <img
                className={Design2ImageViewFamilyImgStyles}
                src={activeSlide.image}
                alt={caption}
                width={aspectRatio.width}
                height={aspectRatio.height}
              />
            </Transition>
            <div className={Design2ImageViewFamilyCaptionContainerStyles}
              data-testid="ImageSlidingGalleryCaptionPart"
            >
              <Caption
                caption={caption}
              />
            </div>
          </div>
          {this.renderPrevNextButtonGroup()}
        </div>
      </div>
    );
  }

  renderPrevNextButtonGroup() {
    return (
      <div className="ImageSlidingGalleryButtonContainer">
        <Button
          icon="arrow alternate circle left outline"
          attached="left"
          size="huge"
          color="blue"
          aria-label="Previous slide"
          disabled={(this.state.activeIndex === 0)}
          onClick={this.handleOnPrevClick}
        />
        <Button
          icon="arrow alternate circle right outline"
          attached="right"
          size="huge"
          color="blue"
          aria-label="Next slide"
          disabled={(this.state.activeIndex === (this.props.slides.length - 1))}
          onClick={this.handleOnNextClick}
        />
      </div>
    );
  }
}
