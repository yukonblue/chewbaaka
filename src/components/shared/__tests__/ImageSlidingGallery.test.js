/**
 * ImageSlidingGallery.test.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 06, 2020
 * Updated  : Aug 23, 2021
 */

import React from 'react'

import { render, screen } from '@testing-library/react'

import renderer from 'react-test-renderer'

import ImageSlidingGallery from '../ImageSlidingGallery'

const slides=[
  {
    image: '',
    caption: "Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words",
    aspectRatio: {
      width: 1600,
      height: 900
    }
  },
  {
    image: '',
    caption: "Listen to a Tsimshian (Pacific Northwest Native) storyteller recount the family history painted on a 38-foot-long house front as specific parts of the design light up.",
    credit: "Image God",
    aspectRatio: {
      width: 1024,
      height: 736
    }
  }
];

test('renders ImageSlidingGallery component', () => {
  render(
    <ImageSlidingGallery
      slides={slides}
    />
  );

  const captionElement = screen.getByTestId("ImageSlidingGalleryCaptionPart");
  expect(captionElement).toBeInTheDocument();
  expect(captionElement.textContent).toBe(slides[0].caption);
});

test('ImageSlidingGallery component snapshot', () => {
  const tree = renderer
    .create(
      <ImageSlidingGallery
        slides={slides}
      />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
