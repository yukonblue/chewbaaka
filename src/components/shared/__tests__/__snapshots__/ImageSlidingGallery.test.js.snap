// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlidingGallery component snapshot 1`] = `
<div
  className="ImageSlidingGalleryOuterContainer"
  style={
    Object {
      "maxWidth": undefined,
    }
  }
>
  <div
    className="ImageSlidingGalleryCoreContainer"
  >
    <img
      alt="Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words"
      className="ImageSlidingGalleryImgPart"
      height={900}
      src=""
      width={1600}
    />
    <div
      className="ImageSlidingGalleryButtonContainer"
    >
      <div
        aria-label="Previous slide"
        className="ui blue huge icon left attached disabled button"
        onClick={[Function]}
        role="button"
        tabIndex={-1}
      >
        <i
          aria-hidden="true"
          className="arrow alternate circle left outline icon"
          onClick={[Function]}
        />
      </div>
      <div
        aria-label="Next slide"
        className="ui blue huge icon right attached button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        <i
          aria-hidden="true"
          className="arrow alternate circle right outline icon"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="ImageSlidingGalleryCaptionPart"
    data-testid="ImageSlidingGalleryCaptionPart"
  >
    <div
      className="CaptionContainer"
      data-testid="CaptionComponentTestId"
      style={
        Object {
          "maxWidth": undefined,
        }
      }
    >
      <div
        className="IconContainer"
      >
        <i
          aria-hidden="true"
          className="image icon"
          onClick={[Function]}
        />
      </div>
      <p
        className="ContentPageCaptionTextSize CaptionText"
      >
        Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words
      </p>
    </div>
  </div>
</div>
`;
